### ISSUE TITLE
Add complex password requirements to generate command

### ISSUE THREAD

**Issue by mackenbach** [secrethub/secrethub-cli#177]
Though we discourage users from applying requirements like minimum 1 symbol (as it decreases the entropy of the generated password), we do not control the password requirements of third party applications our users store and generate passwords for. 

And those requirements can be super wacky:

<img src="https://user-images.githubusercontent.com/7395839/65518456-9f6ad800-dee4-11e9-9e57-0560de7787fb.png" width="300px">

To support those password requirements in a relatively simple way, we've come up with three flags:

```
--include  <charset> Include given characters in the set of characters to randomly choose a password from. Defaults to alphanumeric. As the password is randomly chosen, it is not ensured that the password will always contain at least one character from the given character set. Use --min for that. 

--exclude <charset>  Ensure the password does not contain any characters from the given character set. 

--min  <charset>:<n> Ensure that the resulting password contains at least n characters from the given character set.
```

The supported character sets are as follows:

```
alphanumeric    = a-zA-Z0-9
symbols         = !@#$%^*-_+=.,?
similar         = iIlL1oO0
numeric         = 0-9
letters         = a-zA-Z
lowercase       = a-z
uppercase       = A-Z 
```

A few notes on the character sets:
- As `numeric` is a bit less generic, we could have aliases for `numbers`, `number`, and `digits`?
- I think the `similar` set should also include the nine and the lowercase G: `9g`. This requires a change to the `randchar` library though.

A few examples of usage are:

```
$ secrethub generate --include symbols
$ secrethub generate --include symbols --exclude similar
$ secrethub generate --min symbols:2 --min numeric:1
```

Rules are as follows:
- `--include` and `--min` are are mutually exclusive with `--exclude`, meaning you can't have both `--include symbols --exclude symbols` or `--exclude symbols --min symbols:1`.
- `--min symbols` automatically also adds `--include symbols`.
- `--min symbols:0` automatically adds `--include symbols`, but does not impose a minimum constraint

Also, the help text of the `secrethub generate` command should include something about the security implications of constraints. See the package comments of the `randchar` package:

> However, note that the strongest passwords are the ones with
the least requirements. Imposing e.g. a minimum requirement
on generated passwords reduces the entropy of the resulting
password. So when possible, avoid using any constraints.

And finally, the docs texts could use a little work. 

The spec is still a bit of a work in progress, so feedback is welcome.

**Comment by SimonBarendse** [comment]
```
--include  <charset> Include given characters in the set of characters to randomly choose a password from. Defaults to alphanumeric. ...
```

Does a specified `--include` override the default of `alphanumeric`? Or is the used charset the result of combining the default charset with the supplied include and excludes?

Can you pass multiple charsets to include or exclude? e.g. `--include symbols --include alphanumeric` And how about specifying multiple minima?

**Comment by mackenbach** [comment]
> Can you pass multiple charsets to include or exclude? e.g. --include symbols --include alphanumeric And how about specifying multiple minima?

Yes you can specify multiple `--include`, `--exclude` and `--min` flags that will all be combined.

**Comment by mackenbach** [comment]
> Does a specified --include override the default of alphanumeric? Or is the used charset the result of combining the default charset with the supplied include and excludes?

Good question. I'd say the simplest would be to always have `alphanumeric` and then if you add `--include` it includes that with the `alphanumeric`? That would save you from potentially only choosing from symbols if you do `secrethub generate --include symbols`.

And if you want a 'clean slate' you could say something like `secrethub generate --exclude alphanumeric --include symbols` I suppose. 

Not entirely sure what this does with the non-mutually exclusive constraint though. 

What do you think is the most intuitive way to go about this?

**Comment by SimonBarendse** [comment]
> The supported character sets are as follows:
> ```
>alphanumeric    = a-zA-Z0-9
>symbols         = !@#$%^*-_+=.,?
>similar         = iIlL1oO0
>numeric         = 0-9
>letters         = a-zA-Z
>lowercase       = a-z
>uppercase       = A-Z 
>```

> - As numeric is a bit less generic, we could have aliases for numbers, number, and digits?

Would be nice to create a mapping or function upstream in `randchar` of [secrethub-go](https://github.com/secrethub/secrethub-go) that takes the name of a charset (e.g. `"symbols"`) and returns the corresponding `randchar.Charset` (e.g. `randchar.Symbols`). So that we can use the same names consistently across all consumers (e.g. if we add the same functionality to the [Terraform Provider](https://github.com/secrethub/terraform-provider-secrethub).

**Comment by SimonBarendse** [comment]
> I'd say the simplest would be to always have `alphanumeric` and then if you add `--include` it includes that with the `alphanumeric`

This also has my preference.

> And if you want a 'clean slate' you could say something like `secrethub generate --exclude alphanumeric --include symbols` I suppose.
>
>Not entirely sure what this does with the non-mutually exclusive constraint though.

If we want to keep the mutually exclusive constraint (which I think we should), this would mean that if you want to keep part of the `alphanumeric` charset, you should not exclude it entirely, but rather just exclude the parts you don't want, e.g. to have symbols and letters:
```
secrethub generate --include symbols --exclude numbers
```
or to have symbols and lowercase letters:
```
secrethub generate --include symbols --exclude numbers --exclude uppercase
```

**Comment by SimonBarendse** [comment]
Shall we move the default charset explanation from the `--include` help-text to the commands help-text? I think that will prevent the question whether supplying `--include` values overrides the default charset. Also, the default charset is just as relevant for the `--exclude` flag (so it doesn't belong just to `--include`).

**Comment by SimonBarendse** [comment]
As `--include symbols` behaves the same as the current `--symbols` flag, we'll deprecate `--symbols` when implementing `--include`.

@Marton6 When we deprecate flags/args/commands in the CLI, we hide their help text (in new releases), so that the help-text stays clean and it is easy to see for the user what they should use.

**Comment by Marton6** [comment]
@SimonBarendse and I discussed an alternative implementation for this command.

Having the mutual exclusivity constraint can make the command more difficult to use, for example:
```
secrethub generate --exclude alphanumeric --include lowercase
```
would be incorrect and the user would be required to know the default charset to use the command correctly.

Not having this constraint would make the command ambiguous and the result would be dependent on the order of evaluation of the flags.

Instead of the `include` and `exclude` flags, the `--charset` flag could be used for specifying the charset of the password. For example, a password containing only lowercase letters and numbers would be generated like this:
```
secrethub generate --charset=lowercase,numbers
```
The default value of the flag would be `alphanumeric`.

This approach would bring the following advantages:
1. The charset used is clearly visible in the command.
2. The order of the charsets/flags does not matter, there is no ambiguity.
3. There are no constraints (e.g. mutual exclusivity) that could make the command more difficult to use.
4. Only one type of flag is sufficient. (simpler usage and help text)
5. The user is not required to know the default charset.

**Comment by SimonBarendse** [comment]
> Not having this constraint would make the command ambiguous and the result would be dependent on the order of evaluation of the flags.

Depending on the order of evaluation becomes a problem because the configuration isn't always supplied in a context where we have order, e.g.:
- From environment variables
- In configuration languages (e.g. the [generate block of the secret resource](https://github.com/secrethub/terraform-provider-secrethub/blob/v0.3.0/secrethub/resource_secret.go#L46) in the Terraform provider),

**Comment by mackenbach** [comment]
Sounds good to me. Smart move to make it simpler. A few questions to consider:

- So do I understand it correctly that the options for the `--charset` flag will be only basic sets (e.g. `numbers, lowercase, uppercase, symbols`), and not combined ones like `alphanumeric`, `letters`?
- How do you exclude similar characters? Is this going to be a separate flag?
- Does this decision impact how the user specifies a minimum number of characters from one set (with the `--min` flag)?

**Comment by SimonBarendse** [comment]
Good questions! :+1: 

> - So do I understand it correctly that the options for the `--charset` flag will be only basic sets (e.g. `numbers, lowercase, uppercase, symbols`), and not combined ones like `alphanumeric`, `letters`?

I think there's still value in keeping the combined charsets as shorthands:
`--charset alphanumeric` and `--charset lowercase,uppercase,numbers` are equivalent, but the latter is faster to type and read. Same goes for e.g.
`--charset alphanumeric,symbols` and `--charset lowercase,uppercase,numbers,symbols`

> - How do you exclude similar characters? Is this going to be a separate flag?
In the new design the idea is to not include characters you don't want, instead of excluding them.

I think that'd mean creating a charset that includes all alphanumeric but similar characters. Naming might be though though. How about `human-readable`, `no-similar` or `alphanumeric-no-similar`?

Usage: `--charset alphanumeric-no-similar,symbols`

> - Does this decision impact how the user specifies a minimum number of characters from one set (with the `--min` flag)?

No. If a minimum number of characters of a given set is required, this is still supplied with `--min <charset>:<n>`.

We might want to drop this proposed behavior though: `--min symbols automatically also adds --include symbols.` and error instead if no character from the charset is included (= the requirements cannot all be satisfied).

**Comment by mackenbach** [comment]
>> How do you exclude similar characters? Is this going to be a separate flag?
> In the new design the idea is to not include characters you don't want, instead of excluding them.
> 
> I think that'd mean creating a charset that includes all alphanumeric but similar characters. Naming might be tough though. How about human-readable, no-similar or alphanumeric-no-similar?

I feel we're creating more problems than we're solving now. This would require us to add a `xxx-no-similar` shorthand for every charset we have. If that is the case, I'd prefer a `--exclude-similar` flag, but then again that isn't forward compatible if we ever want to exclude other characters from the set.

**Comment by SimonBarendse** [comment]
> This would require us to add a xxx-no-similar shorthand for every charset we have.

Do we? Is there a use case for all of these options? Are they common enough that we want to provide a shorthand for that?

What if we would add the possibility to use a custom charset? Then you can make all options, include the other options for which we don't provide a shorthand: e.g. all letters from a-q, hexadecimal chars etc.

**Comment by mackenbach** [comment]
Okay, I feel like we keep having very slow cycles of back and forth. Let's have this discussion quickly in person to hash it out and decide what to do.

**Comment by SimonBarendse** [comment]
We concluded to go with the `human-readable` charset that will contain `alphanumeric`-`similar` charset. To support other charsets, we eventually want to add the possibility to use a custom charset.

**Comment by SimonBarendse** [comment]
We also decided to error if `--min` mentions a charset that is not included in `--charset`.

**Comment by SimonBarendse** [comment]
We also discussed `--min symbols:1` and `--min symbols:2` can both be met, so using those both is equivalent to using `--min symbols:2`.

**Comment by SimonBarendse** [comment]
This is implemented in https://github.com/secrethub/secrethub-cli/pull/236 and will be included in the next release.

**Comment by SimonBarendse** [comment]
This is released in [v0.37.0](https://github.com/secrethub/secrethub-cli/releases/tag/v0.37.0)
